salt/orch/migration.sls

```diff
@@ -0,0 +1,85 @@
+{#- Get a list of nodes seem to be down or unresponsive #}
+{#- This sends a "are you still there?" message to all #}
+{#- the nodes and wait for a response, so it takes some time. #}
+
+{%- set parallelism = salt['pillar.get']('parallelism', 3) %}
+
+{%- set nodes_down = salt.saltutil.runner('manage.down') %}
+{%- if nodes_down|length >= 1 %}
+# {{ nodes_down|join(',') }} seem to be down: skipping
+  {%- do salt.caasp_log.debug('CaaS: nodes "%s" seem to be down: ignored', nodes_down|join(',')) %}
+  {%- set is_responsive_node_tgt = 'not L@' + nodes_down|join(',') %}
+{%- else %}
+# all nodes seem to be up
+  {%- do salt.caasp_log.debug('CaaS: all nodes seem to be up') %}
+  {#- we cannot leave this empty (it would produce many " and <empty>" targets) #}
+  {%- set is_responsive_node_tgt = '*' %}
+{%- endif %}
+
+{%- set is_migratable = is_responsive_node_tgt + ' and G@tx_update_migration_available:true' %}
+
+{%- set is_admin = is_migratable + ' and G@roles:admin' %}
+
+{%- set nodes_up = salt.saltutil.runner('manage.up')|reject("equalto", "ca") %}
+
+{% for node in nodes_up %}
+  # Maybe using the cache here might lead to problems, if it is out of date.
+  {%- set migration_grain = salt['saltutil.runner']('cache.grains', tgt=node)[node].get('tx_update_migration_available', false) %}
+  {%- do salt.caasp_log.debug('Checking {0} for migration: {1}'.format(node, migration_grain)) %}
+  {% if not migration_grain %}
+    {% do salt.test.exception('Not all nodes are migratable: {0} is missing'.format(node)) %}
```

Comment:
i had the issue that when the admin was already migrated, the exception here got thrown. especially when we retry the node migration we would always run into this or?

```
admin_master:
    Data failed to compile:
----------
    Rendering SLS 'base:orch.migration' failed: Jinja error: Not all nodes are migratable: admin is missing
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/utils/templates.py", line 368, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 989, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 754, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 30, in top-level template code
  File "/usr/lib/python2.7/site-packages/salt/modules/test.py", line 512, in exception
    raise Exception(message)
Exception: Not all nodes are migratable: admin is missing

; line 30

---
[...]
{% for node in nodes_up %}
  # Maybe using the cache here might lead to problems, if it is out of date.
  {%- set migration_grain = salt['saltutil.runner']('cache.grains', tgt=node)[node].get('tx_update_migration_available', false) %}
  {%- do salt.caasp_log.debug('Checking {0} for migration: {1}'.format(node, migration_grain)) %}
  {% if not migration_grain %}
    {% do salt.test.exception('Not all nodes are migratable: {0} is missing'.format(node)) %}    <======================
  {% endif %}
{% endfor %}

# We have to disable the transactional-update.timer to prevent interference with
# our update. The timer could create a newer snapshot otherwise and break the
[...]
---
```

Reply:
Given that all the `migration` orchestration does is install the updates via `transactional-update` I don't think we should ever run into this situation.

Once this migration has completed (and the grain was removed -> all nodes have the `transactional-update` finished) this one should not be run again, so only the `update` orchestration should be run from now on.